index.js

```diff
@@ -35,37 +35,53 @@ function findOrCreatePlugin(schema, options) {
         });
       }
     }
-    this.findOne(conditions, function(err, result) {
-      if (err || result) {
-        if (options && options.upsert && !err) {
-          self.update(conditions, doc, function(err, count) {
-            self.findById(result._id, function(err, result) {
-              callback(err, result, false);
-            });
-          });
-        } else {
-          callback(err, result, false);
-        }
-      } else {
-        for (var key in doc) {
-         conditions[key] = doc[key];
+    //mongoose 7.0.x does not support callbacks so we use findOne().exec().then().catch()
+    this.findOne(conditions).exec().then(function(result){
+      if(result == null){
+        for(var key in doc){
+          conditions[key] = doc[key];
         }
-
-        // If the value contain `$` remove the key value pair
         var keys = Object.keys(conditions);
-
-        for (var z = 0; z < keys.length; z++) {
+        for(var z = 0;z < keys.length; z++){
           var value = JSON.stringify(conditions[keys[z]]);
-          if (value && value.indexOf('$') !== -1) {
+          if(value && value.indexOf('$') !== -1){
             delete conditions[keys[z]];
           }
         }
-
         var obj = new self(conditions);
-        obj.save(function(err) {
-          callback(err, obj, true);
+        obj.save().then(function(result){
+          err = null;
+          callback(err,obj,true);
+        }).catch(function(err){
+          console.log(err);
```

Comment:
This is wrong. You need to call the `callback` when there is an error and pass the `err` object to it.

Additionally, you shouldn’t add `console.log` code. That may be useful during debugging, but should never be present in production code. Instead, make sure you bubble errors back out to the client of the library (by calling the callback) so that consumers of the library can handle the error in the way that they wish (which may be by logging it, possibly to some other place than STDOUT, but may also be something completely different).

If necessary, you may need to run the testsuite or add tests to make sure more of your newly introduced code branches are exercised and functional. Have you run the testsuite?